operator/v1/types_network.go

```diff
@@ -440,6 +440,10 @@ type OVNKubernetesConfig struct {
 	// cluster.
 	// +optional
 	IPsecConfig *IPsecConfig `json:"ipsecConfig,omitempty"`
+	// ipsecExternal enables external (NS) IPsec for the node, the configuration
+	// is done via nmstate operator.
+	// +optional
+	IPsecExternal *IPsecExternal `json:"ipsecExternal,omitempty"`
```

Comment:
You're using the presence of the struct to enable the external ipsec? Would prefer to have an enum with the struct with an enum defining the state.

Does this field have any interactions with other fields within this configuration? Eg must I not specify some other fields when this field is specified?

Given you're adding a struct, do you expect there to be additional configuration in the future for this feature?

Reply:
This is following the exact same pattern as the above `IPsecConfig`..

there is no interaction with other fields at the API level. the implementation will take care of all the relevant cases, mainly around the interaction between IPsecConfig and IPsecExternal (but they are both independent of each other on the API level)

and yes, it gives us the opportunity to add more config options in the future, without changing the API
see my comment - https://github.com/openshift/api/pull/1667#discussion_r1395917150

Reply:
1 - I use the same mechanism for enable/disable (presence of a struct) as the existing IPsecConfig parameter which is used to enable/disable EW ipsec (although the name suggests differently the configuration is done in other place)
2 - no interaction with other fields, it is completely independent. 
@yuvalk - Does it make sense to put in this struct (in the future) reference to nmstate configuration to be used for the ipsec configuration. (name of the nmstate CRD?)

Reply:
> This is following the exact same pattern as the above IPsecConfig..

That API was merged 3 years ago, and API guidance has changed over time as we learn about prior mistakes. The `Network` resource is considered to be a configuration API. Where we prefer discoverability. To enable discoverability, we implement a few rules, eg, avoiding pointers and omitempty where possible.

Current guidance would suggest that this field should be
```
IPSecExternal IPSecExternal `json:"ipsecExternal"`
```
And the struct should contain a field
```
// +kubebuilder:validation:Enum="";Enabled;Disabled
// +kubebuilder:validation:Required
State IPSecExteranlState `json:"state"`
```

That said, looking at this, why is `IPSecConfig` and `IPSecExternal` not mutually exclusive.

> there is no interaction with other fields at the API level. the implementation will take care of all the relevant cases, mainly around the interaction between IPsecConfig and IPsecExternal (but they are both independent of each other on the API level)

I'm interested in the behaviour that the operator will take care of, not just the interactions at the API level that we have defined so far. Given my question about not being mutually exclusive, I think understanding what the operator does might help here.

Reply:
the names might be misleading
IPSecConfig - the old one. is used to enable/disable East-West IPsec
IPSecExternal - the new one, is used to enable/disable North-South IPsec

both will first deploy a MC to enable the host ipsec extension (ie - install libreswan and enable ipsec.service)
E-W is also running a DS that runs ovs ipsec monitor (used to configre ipsec policies between nodes)
N-S wont do anything else (beside the mentioned MC). ie - configuration is on the customer

there's also logic that handles upgrades etc.

Reply:
Do you think we could make that clearer in the godoc, since I bet most people won't understand. How would you explain the north-south and east-west in laymans terms for the godoc? I think we probably want to update the documentation on both fields here to make it obvious where the traffic goes based on these two ipsec configurations

Reply:
@JoelSpeed @yuvalk I am going to fix this, but I wonder if making this mandatory is not creating bw compatibility so that exiting yaml files will not work anymore. Shouldn't we make it optional with disabled value and remove the omitempty for discoverability?

Reply:
> IPSecExternal IPSecExternal `json:"ipsecExternal"`

@JoelSpeed does this means there is no default and it'll have to be specified at install time? (ie installer also need to change, and any place where even customer is injecting network config)

Reply:
In situations such as this, we normally create an alias for the empty string, so that the empty string has the same meaning, in this case, as disabled.

Reply:
@JoelSpeed - I made all the changes and tested them as much as I can, I hope it is OK now.